File: web/src/lib/metadata_model/metadata_model.ts
export enum FilterConditionLogicalOperators {
	AND = 'AND',
	OR = 'OR',
	AND_NOT = 'AND NOT',
	OR_NOT = 'OR NOT'
}

export interface IDatabaseColumnFields {
	column_fields_read_order: string[]
	fields: { [key: string]: IMetadataModel }
}

export type QueryConditions = { [key: string]: IQueryCondition }

export interface ISearchResults {
	metadata_model?: IMetadataModel
	data?: any[]
}

export interface IQueryCondition {
	[QcProperties.FG_PROPERTY]?: IMetadataModel
	[QcProperties.FG_FILTER_CONDITION]?: IFilterCondition[][]
	[QcProperties.D_SORT_BY_ASC]?: boolean
	[QcProperties.D_FULL_TEXT_SEARCH_QUERY]?: string
}

export enum QcProperties {
	FG_PROPERTY = '$FG_PROPERTY',
	FG_FILTER_CONDITION = '$FG_FILTER_CONDITION',
	D_SORT_BY_ASC = '$D_SORT_BY_ASC',
	D_FULL_TEXT_SEARCH_QUERY = '$D_FULL_TEXT_SEARCH_QUERY'
}

export interface IFilterCondition {
	[FConditionProperties.FILTER_NEGATE]?: boolean
	[FConditionProperties.FILTER_CONDITION]?: FilterCondition
	[FConditionProperties.FILTER_VALUE]?: any[] | any
}

export enum FEqualToValue {
	FILTER_NEGATE = '$FILTER_NEGATE',
	FILTER_CONDITION = '$FILTER_CONDITION',
	FILTER_VALUE = '$FILTER_VALUE'
}

export interface IFConditionFilterEqualToValue {
	[FSelectProperties.TYPE]?: FSelectType
	[FSelectProperties.VALUE]?: any
}

export enum FConditionProperties {
	FILTER_NEGATE = '$FILTER_NEGATE',
	FILTER_CONDITION = '$FILTER_CONDITION',
	FILTER_VALUE = '$FILTER_VALUE'
}

export enum FilterCondition {
	FIELD_GROUP_NO_OF_ENTRIES_GREATER_THAN = 'FILTER_FG_NO_OF_ENTRIES_GREATER_THAN',
	FIELD_GROUP_NO_OF_ENTRIES_LESS_THAN = 'FILTER_FG_NO_OF_ENTRIES_LESS_THAN',
	FIELD_GROUP_NO_OF_ENTRIES_EQUAL_TO = 'FILTER_FG_NO_OF_ENTRIES_EQUAL_TO',
	FIELD_NUMBER_GREATER_THAN = 'FILTER_F_NUMBER_GREATER_THAN',
	FIELD_NUMBER_LESS_THAN = 'FILTER_F_NUMBER_LESS_THAN',
	FIELD_TIMESTAMP_GREATER_THAN = 'FILTER_F_TIMESTAMP_GREATER_THAN',
	FIELD_TIMESTAMP_LESS_THAN = 'FILTER_F_TIMESTAMP_LESS_THAN',
	FIELD_EQUAL_TO = 'FILTER_F_EQUAL_TO',
	FIELD_TEXT_BEGINS_WITH = 'FILTER_F_TEXT_BEGINS_WITH',
	FIELD_TEXT_ENDS_WITH = 'FILTER_F_TEXT_ENDS_WITH',
	FIELD_TEXT_CONTAINS = 'FILTER_F_TEXT_CONTAINS'
}

export const EmptyMetadataModel = () =>
	({
		[FgProperties.FIELD_GROUP_KEY]: '$',
		[FgProperties.FIELD_GROUP_NAME]: 'root',
		[FgProperties.FIELD_GROUP_DESCRIPTION]: 'group description',
		[FgProperties.FIELD_GROUP_MAX_ENTRIES]: 1,
		[FgProperties.GROUP_FIELDS]: [{}],
		[FgProperties.GROUP_READ_ORDER_OF_FIELDS]: []
	}) as IMetadataModel

export interface IMetadataModel {
	[FgProperties.FIELD_GROUP_KEY]: string
	[FgProperties.FIELD_GROUP_NAME]?: string
	[FgProperties.FIELD_GROUP_DESCRIPTION]?: string
	[FgProperties.FIELD_GROUP_VIEW_TABLE_LOCK_COLUMN]?: boolean
	[FgProperties.GROUP_VIEW_TABLE_IN_2D]?: boolean
	[FgProperties.GROUP_QUERY_ADD_FULL_TEXT_SEARCH_BOX]?: boolean
	[FgProperties.FIELD_GROUP_IS_PRIMARY_KEY]?: boolean
	[FgProperties.FIELD_DATATYPE]?: FieldType
	[FgProperties.FIELD_UI]?: FieldUi
	[FgProperties.FIELD_GROUP_VIEW_VALUES_IN_SEPARATE_COLUMNS]?: boolean
	[FgProperties.FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS]?: number
	[FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT]?: string
	[FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX]?: number
	[FgProperties.FIELD_DATETIME_FORMAT]?: FieldDateTimeFormat
	[FgProperties.FIELD_SELECT_OPTIONS]?: ISelectOption[]
	[FgProperties.FIELD_PLACEHOLDER]?: string
	[FgProperties.FIELD_GROUP_MAX_ENTRIES]?: number
	[FgProperties.FIELD_DEFAULT_VALUE]?: any
	[FgProperties.FIELD_GROUP_INPUT_DISABLE]?: boolean
	[FgProperties.FIELD_GROUP_DISABLE_PROPERTIES_EDIT]?: boolean
	[FgProperties.FIELD_CHECKBOX_VALUE_IF_TRUE]?: IFieldCheckboxValue
	[FgProperties.FIELD_CHECKBOX_VALUE_IF_FALSE]?: IFieldCheckboxValue
	[FgProperties.FIELD_CHECKBOX_VALUES_USE_IN_VIEW]?: boolean
	[FgProperties.FIELD_CHECKBOX_VALUES_USE_IN_STORAGE]?: boolean
	[FgProperties.FIELD_GROUP_VIEW_DISABLE]?: boolean
	[FgProperties.FIELD_GROUP_QUERY_CONDITIONS_EDIT_DISABLE]?: boolean
	[FgProperties.GROUP_EXTRACT_AS_SINGLE_FIELD]?: boolean

	[FgProperties.GROUP_READ_ORDER_OF_FIELDS]?: string[]
	[FgProperties.GROUP_FIELDS]?: { [key: string]: IMetadataModel }[]

	[FgProperties.DATABASE_TABLE_COLLECTION_NAME]?: string
	[FgProperties.DATABASE_TABLE_COLLECTION_UID]?: string
	[FgProperties.DATABASE_FIELD_COLUMN_NAME]?: string
	[FgProperties.DATABASE_FIELD_ADD_DATA_TO_FULL_TEXT_SEARCH_INDEX]?: boolean
	[FgProperties.DATABASE_SKIP_DATA_EXTRACTION]?: boolean
	[FgProperties.FIELD_2D_VIEW_POSITION]?: I2DFieldViewPosition
}

export enum FieldType {
	TEXT = 'text',
	NUMBER = 'number',
	BOOLEAN = 'boolean',
	TIMESTAMP = 'timestamp',
	ANY = 'any'
}

export enum FieldUi {
	TEXT = 'text',
	TEXTAREA = 'textarea',
	NUMBER = 'number',
	CHECKBOX = 'checkbox',
	SELECT = 'select',
	DATETIME = 'datetime',
	GROUP = 'group',
	UUID = 'uuid'
}

export enum FieldDateTimeFormat {
	YYYYMMDDHHMM = 'yyyy-mm-dd hh:mm',
	YYYYMMDD = 'yyyy-mm-dd',
	YYYYMM = 'yyyy-mm',
	HHMM = 'hh:mm',
	YYYY = 'yyyy',
	MM = 'mm'
}

export const ARRAY_INDEX_PLACEHOLDER_REGEX_SEARCH = /\[\*\]/

export enum FSelectProperties {
	TYPE = '$TYPE',
	LABEL = '$LABEL',
	VALUE = '$VALUE'
}

export enum FSelectType {
	NUMBER = 'number',
	TEXT = 'text',
	BOOLEAN = 'boolean'
}

export interface ISelectOption {
	[FSelectProperties.LABEL]?: string
	[FSelectProperties.TYPE]?: FSelectType
	[FSelectProperties.VALUE]?: any
}

export enum FieldCheckboxValueProperties {
	TYPE = '$TYPE',
	VALUE = '$VALUE'
}

export interface IFieldCheckboxValue {
	[FieldCheckboxValueProperties.TYPE]?: FieldType
	[FieldCheckboxValueProperties.VALUE]?: any
}

export enum FgProperties {
	FIELD_GROUP_KEY = '$FIELD_GROUP_KEY',
	FIELD_GROUP_NAME = '$FIELD_GROUP_NAME',
	FIELD_GROUP_DESCRIPTION = '$FIELD_GROUP_DESCRIPTION',
	FIELD_GROUP_VIEW_TABLE_LOCK_COLUMN = '$FIELD_GROUP_VIEW_TABLE_LOCK_COLUMN',
	GROUP_VIEW_TABLE_IN_2D = '$GROUP_VIEW_TABLE_IN_2D',
	GROUP_QUERY_ADD_FULL_TEXT_SEARCH_BOX = '$GROUP_QUERY_ADD_FULL_TEXT_SEARCH_BOX',
	FIELD_GROUP_IS_PRIMARY_KEY = '$FIELD_GROUP_IS_PRIMARY_KEY',
	FIELD_DATATYPE = '$FIELD_DATATYPE',
	FIELD_UI = '$FIELD_UI',
	FIELD_GROUP_VIEW_VALUES_IN_SEPARATE_COLUMNS = '$FIELD_GROUP_VIEW_VALUES_IN_SEPARATE_COLUMNS',
	FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS = '$FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS',
	FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT = '$FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT',
	FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX = '$FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX',
	FIELD_INPUT_PLACEHOLDER = '$FIELD_INPUT_PLACEHOLDER',
	FIELD_DATETIME_FORMAT = '$FIELD_DATETIME_FORMAT',
	FIELD_SELECT_OPTIONS = '$FIELD_SELECT_OPTIONS',
	FIELD_PLACEHOLDER = '$FIELD_PLACEHOLDER',
	FIELD_GROUP_MAX_ENTRIES = '$FIELD_GROUP_MAX_ENTRIES',
	FIELD_DEFAULT_VALUE = '$FIELD_DEFAULT_VALUE',
	FIELD_GROUP_INPUT_DISABLE = '$FIELD_GROUP_INPUT_DISABLE',
	FIELD_GROUP_DISABLE_PROPERTIES_EDIT = '$FIELD_GROUP_DISABLE_PROPERTIES_EDIT',
	DATABASE_FIELD_ADD_DATA_TO_FULL_TEXT_SEARCH_INDEX = '$DATABASE_FIELD_ADD_DATA_TO_FULL_TEXT_SEARCH_INDEX',
	FIELD_CHECKBOX_VALUE_IF_TRUE = '$FIELD_CHECKBOX_VALUE_IF_TRUE',
	FIELD_CHECKBOX_VALUE_IF_FALSE = '$FIELD_CHECKBOX_VALUE_IF_FALSE',
	FIELD_CHECKBOX_VALUES_USE_IN_VIEW = '$FIELD_CHECKBOX_VALUES_USE_IN_VIEW',
	FIELD_CHECKBOX_VALUES_USE_IN_STORAGE = '$FIELD_CHECKBOX_VALUES_USE_IN_STORAGE',
	FIELD_GROUP_VIEW_DISABLE = '$FIELD_GROUP_VIEW_DISABLE',
	FIELD_GROUP_QUERY_CONDITIONS_EDIT_DISABLE = '$FIELD_GROUP_QUERY_CONDITIONS_EDIT_DISABLE',
	GROUP_EXTRACT_AS_SINGLE_FIELD = '$GROUP_EXTRACT_AS_SINGLE_FIELD',
	GROUP_READ_ORDER_OF_FIELDS = '$GROUP_READ_ORDER_OF_FIELDS',
	GROUP_FIELDS = '$GROUP_FIELDS',
	DATABASE_SKIP_DATA_EXTRACTION = '$DATABASE_SKIP_DATA_EXTRACTION',
	DATABASE_TABLE_COLLECTION_UID = '$DATABASE_TABLE_COLLECTION_UID',
	DATABASE_TABLE_COLLECTION_NAME = '$DATABASE_TABLE_COLLECTION_NAME',
	DATABASE_FIELD_COLUMN_NAME = '$DATABASE_FIELD_COLUMN_NAME',
	DATUM_INPUT_VIEW = '$DATUM_INPUT_VIEW',
	FIELD_2D_VIEW_POSITION = '$FIELD_2D_VIEW_POSITION'
}

export enum Field2dPositionProperties {
	FIELD_GROUP_KEY = FgProperties.FIELD_GROUP_KEY,
	FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX = FgProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX,
	FIELD_POSITION_BEFORE = '$FIELD_POSITION_BEFORE'
}

export interface I2DFieldViewPosition {
	[Field2dPositionProperties.FIELD_GROUP_KEY]: string
	[Field2dPositionProperties.FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX]?: number
	[Field2dPositionProperties.FIELD_POSITION_BEFORE]?: boolean
}

export enum DView {
	TABLE = 'table',
	FORM = 'form'
}

export function GetNextDataInputView(currentView: DView) {
	switch (currentView) {
		case DView.TABLE:
			return DView.FORM
		default:
			return DView.TABLE
	}
}

export const ARRAY_INDEX_PLACEHOLDER = '[*]'

/**
 * Prepares the path to value in an object based on the metadata-model `$FG_KEY` property of a field in a group.
 *
 * @throws {Error} if:
 * * Number of array index placeholders in {@linkcode fieldgroupkey} being more than the number of array indexes in {@linkcode arrayplaceholderindexes}.
 * * {@linkcode arrayplaceholderindexes} not being an array of numbers ONLY.
 *
 * @param fieldgroupkey MUST begin with `$.$G_FIELDS[*]`. Examples:
 * * `$.$G_FIELDS[*].field_1`.
 * * `$.$G_FIELDS[*].group_1.$G_FIELDS[*].group_1_field`.
 *
 * @param arrayplaceholderindexes Replaces array index placeholder `[*]` found in {@linkcode fieldgroupkey}.
 *
 * Number of elements MUST match number of array index placeholders in {@linkcode fieldgroupkey}.
 * For example, with {@linkcode fieldgroupkey} like `$.$G_FIELDS[*].group_1.$G_FIELDS[*].group_1_field` the number of array indexes passed in {@linkcode arrayplaceholderindexes} MUST be 2.
 *
 * The first element is removed as it matches the first `$G_FIELDS[*]` in the {@linkcode fieldgroupkey} which is removed from the path.
 * For example, {@linkcode fieldgroupkey} `$.$G_FIELDS[*].group_1.$G_FIELDS[*].group_1_field` will be trimmed to `$.group_1[*].group_1_field` before {@linkcode arrayplaceholderindexes} are added.
 *
 * @returns path to value in the object.
 */
export function PreparePathToValueInObject(fieldgroupkey: string, arrayplaceholderindexes: number[] = []): string {
	fieldgroupkey = fieldgroupkey.replace(/\.\$GROUP_FIELDS\[\*\]/, '')
	fieldgroupkey = fieldgroupkey.replace(new RegExp(GROUP_FIELDS_REGEX_SEARCH, 'g'), '')
	arrayplaceholderindexes = arrayplaceholderindexes.slice(1)
	if (!Array.isArray(arrayplaceholderindexes)) {
		throw [PreparePathToValueInObject.name, 'argument groupindexes is not array']
	}
	for (let i = 0; i < arrayplaceholderindexes.length; i++) {
		if (Number.isNaN(arrayplaceholderindexes[i])) {
			throw [PreparePathToValueInObject.name, 'one or more group indexes is not a number']
		}
		fieldgroupkey = fieldgroupkey.replace(ARRAY_PATH_REGEX_SEARCH, `[${arrayplaceholderindexes[i]}]`)
	}
	if (fieldgroupkey.includes(ARRAY_INDEX_PLACEHOLDER)) {
		throw [PreparePathToValueInObject.name, `number of array index placeholders in argument path '${fieldgroupkey}' does not match number of indexes in '${arrayplaceholderindexes.join(',')}'`]
	}
	return fieldgroupkey
}

export function GroupCanBeProcessedAs2D(fieldgroup: any) {
	if (!Array.isArray(fieldgroup[FgProperties.GROUP_FIELDS]) || typeof fieldgroup[FgProperties.GROUP_FIELDS][0] !== 'object' || Array.isArray(fieldgroup[FgProperties.GROUP_FIELDS][0])) {
		return false
	}

	for (const fgValues of Object.values(fieldgroup[FgProperties.GROUP_FIELDS][0])) {
		if (Array.isArray((fgValues as any)[FgProperties.GROUP_FIELDS]) || Array.isArray((fgValues as any)[FgProperties.GROUP_READ_ORDER_OF_FIELDS])) {
			return false
		}
	}

	return true
}

export const ARRAY_PATH_REGEX_SEARCH = /\[\*\]/

export const GROUP_FIELDS_PATH_REGEX_SEARCH = /\$GROUP_FIELDS\[\*\](?:\.|)/

export const GROUP_FIELDS_REGEX_SEARCH = /(?:\.|)\$GROUP_FIELDS/

export const SPECIAL_CHARS_REGEX_SEARCH = /[^a-zA-Z0-9]+/

export const FieldGroupKeyPath = (fieldGroupKey: string) => fieldGroupKey.replace(/\.\$GROUP\_FIELDS\[\*\]\./g, '->')

export const FieldGroupDataJsonPath = (fieldGroupKey: string) => fieldGroupKey.replace(new RegExp(GROUP_FIELDS_PATH_REGEX_SEARCH), '').replace(new RegExp(GROUP_FIELDS_REGEX_SEARCH, 'g'), '')

export interface I2DFields {
	fields: (IMetadataModel | any)[]
	reposition: { [key: number]: I2DFieldViewPosition }
}

export type RepositionFields = { [key: number]: I2DFieldViewPosition }

export const IsGroupFieldsValid = (fg: any) => typeof fg === 'object' && !Array.isArray(fg) && fg !== null
export const IsFieldGroupKeyValid = (fgKey: any) => typeof fgKey === 'string'
export const IsGroupReadOrderOfFieldsValid = (groofv: any) => typeof groofv === 'object' && Array.isArray(groofv)
export const IsFieldAField = (fg: any) => typeof fg[FgProperties.FIELD_DATATYPE] === 'string' && typeof fg[FgProperties.FIELD_UI] === 'string'
export const Is2DFieldViewPositionValid = (fg: any) => typeof fg[FgProperties.FIELD_2D_VIEW_POSITION] === 'object' && !Array.isArray(fg[FgProperties.FIELD_2D_VIEW_POSITION]) && typeof fg[FgProperties.FIELD_2D_VIEW_POSITION][Field2dPositionProperties.FIELD_GROUP_KEY] === 'string'
export function GetFieldGroupName(fg: any, defaultValue: string = '#unnamed') {
	if (fg[FgProperties.FIELD_GROUP_NAME]) {
		return fg[FgProperties.FIELD_GROUP_NAME]
	}
	if (typeof fg[FgProperties.FIELD_GROUP_KEY] === 'string') {
		return (fg[FgProperties.FIELD_GROUP_KEY] as string).split('.').pop()
	}

	return defaultValue
}

export type Error = any[]

File: internal/pkg/metadata_model/metadata_model.go
package metadatamodel

import (
	"errors"
	"fmt"
	"reflect"
	"regexp"
	"strings"
)

const (
	QUERY_CONDITION_PROP_FG_PROPERTY              string = "$FG_PROPERTY"
	QUERY_CONDITION_PROP_FG_FILTER_CONDITION      string = "$FG_FILTER_CONDITION"
	QUERY_CONDITION_PROP_D_SORT_BY_ASC            string = "$D_SORT_BY_ASC"
	QUERY_CONDITION_PROP_D_FULL_TEXT_SEARCH_QUERY string = "$D_FULL_TEXT_SEARCH_QUERY"
)

const (
	FILTER_CONDITION_PROP_FILTER_NEGATE    string = "$FILTER_NEGATE"
	FILTER_CONDITION_PROP_FILTER_CONDITION string = "$FILTER_CONDITION"
	FILTER_CONDITION_PROP_FILTER_VALUE     string = "$FILTER_VALUE"
)

const (
	FILTER_CONDTION_NO_OF_ENTRIES_GREATER_THAN string = "NO_OF_ENTRIES_GREATER_THAN"
	FILTER_CONDTION_NO_OF_ENTRIES_LESS_THAN    string = "NO_OF_ENTRIES_LESS_THAN"
	FILTER_CONDTION_NO_OF_ENTRIES_EQUAL_TO     string = "NO_OF_ENTRIES_EQUAL_TO"
	FILTER_CONDTION_NUMBER_GREATER_THAN        string = "NUMBER_GREATER_THAN"
	FILTER_CONDTION_NUMBER_LESS_THAN           string = "NUMBER_LESS_THAN"
	FILTER_CONDTION_TIMESTAMP_GREATER_THAN     string = "TIMESTAMP_GREATER_THAN"
	FILTER_CONDTION_TIMESTAMP_LESS_THAN        string = "TIMESTAMP_LESS_THAN"
	FILTER_CONDTION_EQUAL_TO                   string = "EQUAL_TO"
	FILTER_CONDTION_TEXT_BEGINS_WITH           string = "TEXT_BEGINS_WITH"
	FILTER_CONDTION_TEXT_ENDS_WITH             string = "TEXT_ENDS_WITH"
	FILTER_CONDTION_TEXT_CONTAINS              string = "TEXT_CONTAINS"
)

const (
	FIELD_TYPE_TEXT      string = "text"
	FIELD_TYPE_NUMBER    string = "number"
	FIELD_TYPE_BOOLEAN   string = "boolean"
	FIELD_TYPE_TIMESTAMP string = "timestamp"
	FIELD_TYPE_ANY       string = "any"
)

const (
	FIELD_UI_TEXT     string = "text"
	FIELD_UI_TEXTAREA string = "textarea"
	FIELD_UI_NUMBER   string = "number"
	FIELD_UI_CHECKBOX string = "checkbox"
	FIELD_UI_SELECT   string = "select"
	FIELD_UI_DATETIME string = "datetime"
	FIELD_UI_GROUP    string = "group"
)

const (
	FIELD_DATE_TIME_FORMAT_YYYYMMDDHHMM string = "yyyy-mm-dd hh:mm"
	FIELD_DATE_TIME_FORMAT_YYYYMMDD     string = "yyyy-mm-dd"
	FIELD_DATE_TIME_FORMAT_YYYYMM       string = "yyyy-mm"
	FIELD_DATE_TIME_FORMAT_HHMM         string = "hh:mm"
	FIELD_DATE_TIME_FORMAT_YYYY         string = "yyyy"
	FIELD_DATE_TIME_FORMAT_MM           string = "mm"
)

const (
	FIELD_SELECT_PROP_TYPE  string = "$TYPE"
	FIELD_SELECT_PROP_LABEL string = "$LABEL"
	FIELD_SELECT_PROP_VALUE string = "$VALUE"
)

const (
	FIELD_SELECT_TYPE_NUMBER  string = "number"
	FIELD_SELECT_TYPE_TEXT    string = "text"
	FIELD_SELECT_TYPE_BOOLEAN string = "boolean"
)

const (
	FIELD_CHECKBOX_VALUE_PROP_TYPE  string = "$TYPE"
	FIELD_CHECKBOX_VALUE_PROP_VALUE string = "$VALUE"
)

const (
	FIELD_GROUP_PROP_FIELD_GROUP_KEY                                       string = "$FIELD_GROUP_KEY"
	FIELD_GROUP_PROP_FIELD_GROUP_NAME                                      string = "$FIELD_GROUP_NAME"
	FIELD_GROUP_PROP_FIELD_GROUP_DESCRIPTION                               string = "$FIELD_GROUP_DESCRIPTION"
	FIELD_GROUP_PROP_FIELD_GROUP_VIEW_TABLE_LOCK_COLUMN                    string = "$FIELD_GROUP_VIEW_TABLE_LOCK_COLUMN"
	FIELD_GROUP_PROP_GROUP_VIEW_TABLE_IN_2D                                string = "$GROUP_VIEW_TABLE_IN_2D"
	FIELD_GROUP_PROP_GROUP_QUERY_ADD_FULL_TEXT_SEARCH_BOX                  string = "$GROUP_QUERY_ADD_FULL_TEXT_SEARCH_BOX"
	FIELD_GROUP_PROP_FIELD_GROUP_IS_PRIMARY_KEY                            string = "$FIELD_GROUP_IS_PRIMARY_KEY"
	FIELD_GROUP_PROP_FIELD_DATATYPE                                        string = "$FIELD_DATATYPE"
	FIELD_GROUP_PROP_FIELD_UI                                              string = "$FIELD_UI"
	FIELD_GROUP_PROP_FIELD_GROUP_VIEW_VALUES_IN_SEPARATE_COLUMNS           string = "$FIELD_GROUP_VIEW_VALUES_IN_SEPARATE_COLUMNS"
	FIELD_GROUP_PROP_FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS string = "$FIELD_GROUP_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS"
	FIELD_GROUP_PROP_FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT   string = "$FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT"
	FIELD_GROUP_PROP_FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX    string = "$FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX"
	FIELD_GROUP_PROP_FIELD_INPUT_PLACEHOLDER                               string = "$FIELD_INPUT_PLACEHOLDER"
	FIELD_GROUP_PROP_FIELD_DATETIME_FORMAT                                 string = "$FIELD_DATETIME_FORMAT"
	FIELD_GROUP_PROP_FIELD_SELECT_OPTIONS                                  string = "$FIELD_SELECT_OPTIONS"
	FIELD_GROUP_PROP_FIELD_PLACEHOLDER                                     string = "$FIELD_PLACEHOLDER"
	FIELD_GROUP_PROP_FIELD_GROUP_MAX_ENTRIES                               string = "$FIELD_GROUP_MAX_ENTRIES"
	FIELD_GROUP_PROP_FIELD_DEFAULT_VALUE                                   string = "$FIELD_DEFAULT_VALUE"
	FIELD_GROUP_PROP_FIELD_GROUP_INPUT_DISABLE                             string = "$FIELD_GROUP_INPUT_DISABLE"
	FIELD_GROUP_PROP_FIELD_GROUP_DISABLE_PROPERTIES_EDIT                   string = "$FIELD_GROUP_DISABLE_PROPERTIES_EDIT"
	FIELD_GROUP_PROP_DATABASE_FIELD_ADD_DATA_TO_FULL_TEXT_SEARCH_INDEX     string = "$DATABASE_FIELD_ADD_DATA_TO_FULL_TEXT_SEARCH_INDEX"
	FIELD_GROUP_PROP_FIELD_CHECKBOX_VALUE_IF_TRUE                          string = "$FIELD_CHECKBOX_VALUE_IF_TRUE"
	FIELD_GROUP_PROP_FIELD_CHECKBOX_VALUE_IF_FALSE                         string = "$FIELD_CHECKBOX_VALUE_IF_FALSE"
	FIELD_GROUP_PROP_FIELD_CHECKBOX_VALUES_USE_IN_VIEW                     string = "$FIELD_CHECKBOX_VALUES_USE_IN_VIEW"
	FIELD_GROUP_PROP_FIELD_CHECKBOX_VALUES_USE_IN_STORAGE                  string = "$FIELD_CHECKBOX_VALUES_USE_IN_STORAGE"
	FIELD_GROUP_PROP_FIELD_GROUP_VIEW_DISABLE                              string = "$FIELD_GROUP_VIEW_DISABLE"
	FIELD_GROUP_PROP_FIELD_GROUP_QUERY_CONDITIONS_EDIT_DISABLE             string = "$FIELD_GROUP_QUERY_CONDITIONS_EDIT_DISABLE"
	FIELD_GROUP_PROP_GROUP_EXTRACT_AS_SINGLE_FIELD                         string = "$GROUP_EXTRACT_AS_SINGLE_FIELD"
	FIELD_GROUP_PROP_GROUP_READ_ORDER_OF_FIELDS                            string = "$GROUP_READ_ORDER_OF_FIELDS"
	FIELD_GROUP_PROP_GROUP_FIELDS                                          string = "$GROUP_FIELDS"
	FIELD_GROUP_PROP_DATABASE_SKIP_DATA_EXTRACTION                         string = "$DATABASE_SKIP_DATA_EXTRACTION"
	FIELD_GROUP_PROP_DATABASE_TABLE_COLLECTION_UID                         string = "$DATABASE_TABLE_COLLECTION_UID"
	FIELD_GROUP_PROP_DATABASE_TABLE_COLLECTION_NAME                        string = "$DATABASE_TABLE_COLLECTION_NAME"
	FIELD_GROUP_PROP_DATABASE_FIELD_COLUMN_NAME                            string = "$DATABASE_FIELD_COLUMN_NAME"
	FIELD_GROUP_PROP_DATUM_INPUT_VIEW                                      string = "$DATUM_INPUT_VIEW"
	FIELD_GROUP_PROP_FIELD_2D_VIEW_POSITION                                string = "$FIELD_2D_VIEW_POSITION"
)

const (
	FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY                                    string = FIELD_GROUP_PROP_FIELD_GROUP_KEY
	FIELD_2D_POSITION_PROP_FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX string = FIELD_GROUP_PROP_FIELD_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX
	FIELD_2D_POSITION_PROP_                                                   string = "$FIELD_POSITION_BEFORE"
)

const (
	ARRAY_PATH_PLACEHOLDER string = "[*]"
)

type MetadataModelSearchResults struct {
	MetadataModel any `json:"metadata_model"`
	Data          any `json:"data"`
}

var ErrPathContainsIndexPlaceHolders = errors.New("PathContainsIndexPlaceHolders")

// Prepares the path to value in an object based on the metadatamodel `$FG_KEY`(OBJECT_KEY_FIELD_GROUP_KEY) property of a field in a group.
//
// Parameters:
//
//   - path - path to value in object. Must begin with `$.$G_FIELDS[*]`.
//     Examples: `$.$G_FIELDS[*].field_1` results in `field_1` and `$.$G_FIELDS[*].group_1.$G_FIELDS[*].group_1_field` results in `group_1[*].group_1_field`.
//
//   - groupIndexes - Each element replaces array index placeholder (ARRAY_PATH_REGEX_SEARCH) `[*]` found in path.
//
//     Must NOT be empty as the first element in groupIndexes removed as it matches the first `$G_FIELDS[*]` in the path which is removed from the path since it indicates the root of the metadata-model.
//
//     Number of elements MUST match number of array index placeholders in path.
//
//     For example, with path like `$.$G_FIELDS[*].group_1.$G_FIELDS[*].group_1_field` the number of array indexes passed in groupIndexes MUST be 2.
//
// The first element in groupIndexes removed as it matches the first `$G_FIELDS[*]` in the path which is removed from the path since it indicates the root of the metadata-model.
//
// For example, path `$.$G_FIELDS[*].group_1.$G_FIELDS[*].group_1_field` will be trimmed to `$.group_1[*].group_1_field` before groupIndexes are added.
//
// Return path to value in object or error if the number of array index placeholders in path being more than the number of array indexes in groupIndexes.
func PreparePathToValueInObject(path string, groupIndexes []int) (string, error) {
	path = strings.Replace(path, ".$G_FIELDS[*]", "", 1)
	path = string(G_FIELDS_REGEX_SEARCH().ReplaceAll([]byte(path), []byte("")))
	groupIndexes = groupIndexes[1:]
	for _, groupIndex := range groupIndexes {
		path = strings.Replace(path, ARRAY_PATH_PLACEHOLDER, fmt.Sprintf("[%v]", groupIndex), 1)
		groupIndexes = groupIndexes[1:]
	}

	if strings.Contains(path, ARRAY_PATH_PLACEHOLDER) {
		return path, ErrPathContainsIndexPlaceHolders
	}

	return path, nil
}

func GroupCanBeProcessAs2D(fieldgroup any) bool {
	if fieldGroupMap, ok := fieldgroup.(map[string]any); ok {
		if groupFieldsArray, ok := fieldGroupMap[FIELD_GROUP_PROP_GROUP_FIELDS].([]any); ok && len(groupFieldsArray) == 1 {
			if groupFieldsMap, ok := groupFieldsArray[0].(map[string]any); ok {
				for _, value := range groupFieldsMap {
					if valueMap, ok := value.(map[string]any); ok {
						if reflect.TypeOf(valueMap[FIELD_GROUP_PROP_GROUP_READ_ORDER_OF_FIELDS]).Kind() == reflect.Slice {
							return false
						}
					}
				}
			} else {
				return false
			}
		} else {
			return false
		}
	}

	return true
}

func IfKeySuffixMatchesValues(keyToCheck string, valuesToMatch []string) bool {
	for _, value := range valuesToMatch {
		if strings.HasSuffix(keyToCheck, value) {
			return true
		}
	}

	return false
}

func ARRAY_PATH_REGEX_SEARCH() *regexp.Regexp {
	return regexp.MustCompile(`\[\*\]`)
}

func G_FIELDS_PATH_REGEX_SEARCH() *regexp.Regexp {
	return regexp.MustCompile(`\$G_FIELDS\[\*\](?:\.|)`)
}

func G_FIELDS_REGEX_SEARCH() *regexp.Regexp {
	return regexp.MustCompile(`(?:\.|)\$G_FIELDS`)
}

func SPECIAL_CHARS_REGEX_SEARCH() *regexp.Regexp {
	return regexp.MustCompile(`[^a-zA-Z0-9]+`)
}

func IsFieldAField(f any) bool {
	if field, ok := f.(map[string]any); ok {
		return reflect.TypeOf(field[FIELD_GROUP_PROP_FIELD_DATATYPE]).Kind() == reflect.String && reflect.TypeOf(field[FIELD_GROUP_PROP_FIELD_UI]).Kind() == reflect.String
	}

	return false
}

func Is2DFieldViewPositionValid(f any) bool {
	if field, ok := f.(map[string]any); ok {
		if field2Dposition, ok := field[FIELD_GROUP_PROP_FIELD_2D_VIEW_POSITION].(map[string]any); ok {
			return reflect.TypeOf(field2Dposition[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY]).Kind() == reflect.String
		}
	}

	return false
}

func GetFieldGroupName(fg any, defaultValue string) string {
	if fieldGroup, ok := fg.(map[string]any); ok {
		if fieldGroupName, ok := fieldGroup[FIELD_GROUP_PROP_FIELD_GROUP_NAME].(string); ok && len(fieldGroupName) > 0 {
			return fieldGroupName
		}

		if fieldGroupKey, ok := fieldGroup[FIELD_2D_POSITION_PROP_FIELD_GROUP_KEY].(string); ok && len(fieldGroupKey) > 0 {
			fieldGroupKeyParts := strings.Split(fieldGroupKey, ".")
			if len(fieldGroupKeyParts) > 0 {
				return fieldGroupKeyParts[len(fieldGroupKeyParts)-1]
			}
		}
	}

	if len(defaultValue) == 0 {
		return "#unnamed"
	}

	return defaultValue
}

type QueryCondition struct {
	QueryConditionLabel *string             `json:"$QC_LABEL,omitempty"`
	FgProperty          *MetadataModel      `json:"$FG_PROPERTY,omitempty"`
	FgFilterCondition   [][]FilterCondition `json:"$FG_FILTER_CONDITION,omitempty"`
	SortByAsc           map[string]bool     `json:"$D_SORT_BY_ASC,omitempty"`
	SortByDesc          map[string]bool     `json:"$D_SORT_BY_DESC,omitempty"`
	Distinct            *bool               `json:"$D_DISTINCT,omitempty"`
}

type FilterCondition struct {
	FilterNegate    *bool  `json:"$FILTER_NEGATE,omitempty"`
	FilterCondition string `json:"$FILTER_CONDITION,omitempty"`
	FilterValue     []any  `json:"$FILTER_VALUE,omitempty"`
}

type MetadataModel struct {
	FieldGroupKey                                *string `json:"$FG_KEY,omitempty"`
	FieldGroupName                               *string `json:"$FG_NAME,omitempty"`
	FieldGroupDescription                        *string `json:"$FG_DESCRIPTION,omitempty"`
	GroupViewTableIn2D                           *bool   `json:"$G_VIEW_TABLE_IN_2D,omitempty"`
	GroupQueryAddFullTextSearchBox               *bool   `json:"$G_QUERY_ADD_FULL_TEXT_SEARCH_BOX,omitempty"`
	FgIsPrimaryKey                               *bool   `json:"$FG_IS_PRIMARY_KEY,omitempty"`
	FieldDataType                                *string `json:"$F_DATATYPE,omitempty"`
	FieldUi                                      *string `json:"$F_UI,omitempty"`
	FieldGroupViewValuesInSeparateColumns        *bool   `json:"$FG_VIEW_VALUES_IN_SEPARATE_COLUMNS,omitempty"`
	FieldGroupViewMaxNoOfValuesInSeparateColumns *int    `json:"$FG_VIEW_MAX_NO_OF_VALUES_IN_SEPARATE_COLUMNS,omitempty"`
	FieldViewValuesInSeparateColumnsHeaderFormat *string `json:"$F_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_FORMAT,omitempty"`
	FieldViewValuesInSeparateColumnsHeaderIndex  *int    `json:"$F_VIEW_VALUES_IN_SEPARATE_COLUMNS_HEADER_INDEX,omitempty"`
	FieldDatetimeFormat                          *string `json:"$F_DATETIME_FORMAT,omitempty"`
	FieldSelectOptions                           []struct {
		Type  string `json:"$TYPE,omitempty"`
		Label string `json:"$LABEL,omitempty"`
		Value any    `json:"$VALUE,omitempty"`
	} `json:"$F_SELECT_OPTIONS,omitempty"`
	FieldPlaceholder                *string `json:"$F_PLACEHOLDER,omitempty"`
	FieldGroupMaxEntries            *int    `json:"$FG_MAX_ENTRIES,omitempty"`
	FieldDefaultValue               *any    `json:"$F_DEFAULT_VALUE,omitempty"`
	FieldGroupDisableInput          *bool   `json:"$FG_DISABLE_INPUT,omitempty"`
	FieldGroupDisablePropertiesEdit *bool   `json:"$FG_DISABLE_PROPERTIES_EDIT,omitempty"`
	FieldAddToFullTextSearchIndex   *bool   `json:"$F_ADD_TO_FULL_TEXT_SEARCH_INDEX,omitempty"`
	FieldCheckboxValueIfTrue        *any    `json:"$F_CHECKBOX_VALUE_IF_TRUE,omitempty"`
	FieldCheckboxValueIfFalse       *any    `json:"$F_CHECKBOX_VALUE_IF_FALSE,omitempty"`
	FieldCheckboxUseInView          *bool   `json:"$F_CHECKBOX_VALUES_USE_IN_VIEW,omitempty"`
	FieldCheckboxUseInStorage       *bool   `json:"$F_CHECKBOX_VALUES_USE_IN_STORAGE,omitempty"`
	FieldGroupViewDisable           *bool   `json:"$FG_VIEW_DISABLE,omitempty"`
	FieldGroupSkipDataExtraction    *bool   `json:"$FG_SKIP_DATA_EXTRACTION,omitempty"`
	FieldGroupFilterDisable         *bool   `json:"$FG_FILTER_DISABLE,omitempty"`
	GroupExtractAsSingleValue       *bool   `json:"$G_EXTRACT_AS_SINGLE_VALUE,omitempty"`
	MetadataModelGroup
	TableCollectionName *string `json:"$D_TABLE_COLLECTION_NAME,omitempty"`
	FieldColumnName     *string `json:"$D_FIELD_COLUMN_NAME,omitempty"`
}
type MetadataModelGroup struct {
	GroupReadOrderOfFields []string                   `json:"$G_READ_ORDER_OF_FIELDS,omitempty"`
	GroupFieldsIndex       int                        `json:"$G_FIELDS_INDEX,omitempty"`
	GroupFields            []map[string]MetadataModel `json:"$G_FIELDS,omitempty"`
}

type MetadataModelDatabase struct {
	MetadataModel
	TableCollectionName *string `json:"$D_TABLE_COLLECTION_NAME,omitempty"`
	FieldColumnName     *string `json:"$D_FIELD_COLUMN_NAME,omitempty"`
}
